pkg/operator/resource/resourceapply/admissions.go

```diff
@@ -2,76 +2,159 @@ package resourceapply
+	for i, mutatingWebhook := range required.Webhooks {
+		if webhook, ok := existingMutatingWebhooks[mutatingWebhook.Name]; ok {
+			mutatingWebhook.ClientConfig.CABundle = webhook.ClientConfig.CABundle
+		}
+		webhooks[i] = mutatingWebhook
+	}
+	required.Webhooks = webhooks
+}
+
+// ApplyValidatingWebhookConfiguration ensures the form of the specified
+// validatingwebhookconfiguration is present in the API. If it does not exist,
+// it will be created. If it does exist, the metadata of the required
+// validatingwebhookconfiguration will be merged with the existing validatingwebhookconfiguration
+// and an update performed if the validatingwebhookconfiguration spec and metadata differ from
+// the previously required spec and metadata based on generation change.
+//
+// expectedGeneration set to 0 disables generation tracking, forcing redeployment logic.
+func ApplyValidatingWebhookConfiguration(client admissionregistrationclientv1.ValidatingWebhookConfigurationsGetter, recorder events.Recorder,
+	requiredOriginal *admissionregistrationv1.ValidatingWebhookConfiguration, expectedGeneration int64) (*admissionregistrationv1.ValidatingWebhookConfiguration, bool, error) {
+	if requiredOriginal == nil {
+		return nil, false, fmt.Errorf("Unexpected nil instead of an object")
+	}
+	required := requiredOriginal.DeepCopy()
+
+	existing, err := client.ValidatingWebhookConfigurations().Get(context.TODO(), required.GetName(), metav1.GetOptions{})
 	if apierrors.IsNotFound(err) {
-		actual, err := client.MutatingWebhookConfigurations().
-			Create(context.TODO(), required, metav1.CreateOptions{})
+		actual, err := client.ValidatingWebhookConfigurations().Create(context.TODO(), required, metav1.CreateOptions{})
 		reportCreateEvent(recorder, required, err)
-		return actual, true, err
-	}
-	if err != nil {
+		if err != nil {
+			return nil, false, err
+		}
+		return actual, true, nil
+	} else if err != nil {
 		return nil, false, err
 	}
 
 	modified := resourcemerge.BoolPtr(false)
 	existingCopy := existing.DeepCopy()
 
 	resourcemerge.EnsureObjectMeta(modified, &existingCopy.ObjectMeta, required.ObjectMeta)
-	contentSame := equality.Semantic.DeepEqual(existingCopy.Webhooks, required.Webhooks)
-	if contentSame && !*modified {
+	if !*modified && existingCopy.GetGeneration() == expectedGeneration {
 		return existingCopy, false, nil
 	}
-
-	if klog.V(4).Enabled() {
-		klog.Infof("ValidatingWebhookConfiguration %q changes: %v", required.Name, JSONPatchNoError(existing, required))
+	// at this point we know that we're going to perform a write.  We're just trying to get the object correct
+	toWrite := existingCopy // shallow copy so the code reads easier
+	// Providing upgrade compatibility with service-ca-bundle operator
+	// and ignore clientConfig.caBundle changes on "inject-cabundle" label
+	if required.GetAnnotations() != nil && required.GetAnnotations()[genericCABundleInjectorAnnotation] != "" {
```

Comment:
I had a previous comment about this which seem to be gone: why don't we check for a CA cert in required, and keep the existing one if there isn't one enforced. I don't like this dependency on service-ca here. There might be different actors creating CA bundles.

Reply:
Sorry, I hoped it was unresolved. If we do that an the operator just does it's job, then we will get to a hot loop situation with such an approach - https://github.com/openshift/library-go/pull/836/files#r533587142